### DISCUSSION TITLE
iCub Demo

### DISCUSSION THREAD

**Discussion by ivan9a** [robotology/community#513] [Q&A] [answered]
Hi, Where can I download some other demo programs except Redball demo for the iCub? Do you have some demo programs that are ready to be installed ? We soon have a promotion for the public at our faculty, it would be interesting if you have some demo that includes speech, grasping, and gaze control. 

Do you have some instruction how to change the configuration of pf3DTracker in redball demo, pf3DTracker use only picture from left camera to estimate 3D coordinates. I think that 3D estimation would be better if pf3Dtracker use both cameras at the same time, do you have some instructions on how to change that. For now, if we run redball demo we have a significant error between hand and the grasping object. I think that this error is not caused by wrong camera parameters, because I always check parameters before I run pf3Dtracker and redball demo.

Thank you, 
Best Regards,
Ivan Župančić

**Comment by vvasco** [comment, marked as answer]
Hi @ivan9a !
> Where can I download some other demo programs except Redball demo for the iCub? Do you have some demo programs that are ready to be installed ? We soon have a promotion for the public at our faculty, it would be interesting if you have some demo that includes speech, grasping, and gaze control.

You can find a bunch of applications on the [robot bazaar](https://robot-bazaar.iit.it/apps/academy/courses), ready to be downloaded:
- regarding the **speech**, we have [Google speech-to-text](https://robot-bazaar.iit.it/apps/academy/courses/15), [Google Speech Processing](https://robot-bazaar.iit.it/apps/academy/courses/19), and [Google text-to-speech](https://robot-bazaar.iit.it/apps/academy/courses/17), which respectively allow you to convert speech to text, to retrieve the sentence's structure and syntax along with the sentiment, and to generate spoken sentences from written text. You can also find the [Google Speech Application](https://robot-bazaar.iit.it/apps/academy/courses/20) that puts the previous three together. Finally, you can find the [Google Dialog](https://robot-bazaar.iit.it/apps/academy/courses/16) application, that allows you to carry on a conversation about iCub world.
- regarding **grasping** and **gaze control**, at the moment the only application that includes both is [Grasp the Ball Application](https://robot-bazaar.iit.it/apps/academy/courses/1). 

We don't have at the moment an application that includes speech, grasping, and gaze control. However, you might run more applications in parallel for combining them, and eventually, you can develop your own application starting from the ones provided.


>  I think that 3D estimation would be better if pf3Dtracker use both cameras at the same time, do you have some instructions on how to change that. For now, if we run redball demo we have a significant error between hand and the grasping object.I think that this error is not caused by wrong camera parameters, because I always check parameters before I run pf3Dtracker and redball demo.

`pf3Dtracker` works with a single 2D image and extrapolates the 3D position of the ball given that the ball size is known.

Regarding the error you observed, the demo itself provides _reaching offsets_ to adjust the relative position between the ball and the hands. You can either adjust [such offsets](https://github.com/robotology/icub-basic-demos/blob/ba9a10343635741b464abe04cbadb57d2f55eff4/demoRedBall/app/conf/config.ini#L22) manually or rely on the [Grasp the Ball Application](https://robot-bazaar.iit.it/apps/academy/courses/1) which does the procedure automatically if specified. In the latter case, you can follow the instructions at the provided link for downloading and starting the application.

**Reply by ivan9a** [reply]
Hi vvasco,

thank you for your fast response,
regarding the speech i have also tried command: **yarpdev --device speech --lingware-context speech --default-language en --pitch 80 --speed 110** and then i connected two nodes with command **yarp connect /demoRedBall/speech:o /icub/speech:rpc** That works fine for redball demo. I will also try speech application on robot bazzar that you are suggested. 

regarding grasping i tried Grasp the Ball Application on robot bazzar, i have downloaded script and run command but i got following error:
fatal: 
fatal: [icubheadLaptop]: UNREACHABLE! => changed=false 
  msg: |-
    mkdir: cannot create directory '/home/icub/.ansible/tmp/ansible-tmp-1623238831.1588194-276243274592218': No space left on device
  unreachable: true
ok: [icubconsoleLaptop]

I can reach iCub head but still I got error: unreachable, and no space left when I run command **./graspTheBall_setup.sh** 

Best regards,
Ivan Župančić

**Reply by vtikha** [reply]
Hi @ivan9a ,

ok for the **speech**, let us know about the trials with the **speech applications** that @vvasco mentioned on **robot-bazaar**.

As for the second point, were you able to check the space left on the iCub head? You probably need some cleanup or remove some docker images that are on the iCub head?

**Reply by ivan9a** [reply]
Hi vitkha,

Do you have some instructions how to do proper cleanup and how to remove all unnecessary software on iCub head ?

Thank you,
Best regards,
Ivan Zupancic

**Reply by vtikha** [reply]
Hi @ivan9a ,

I would suggest checking the actual space on the **icub-head**. You can display disk usage in a more human-readable format by running the following command:

`df –h`

Have a look at the used/free space of the HD that is used: this is typically `/dev/sda2` and report the findings here.

You can also check the space that `docker` is occupying by running the following commands:

`sudo du -sh /var/lib/docker` and report the findings 

`docker image ls` and report the findings

you can also remove any unwanted images using the following command:

`docker image rm <imageID>`

Please also cleanup the system on the `icub-head` by removing any unwanted exited containers.

`docker container prune`
